src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsMetadataProvider.cs

```diff
@@ -210,7 +251,15 @@ public void CreateDisplayMetadata(DisplayMetadataProviderContext context)
             // TemplateHint
             if (uiHintAttribute != null)
             {
-                displayMetadata.TemplateHint = uiHintAttribute.UIHint;
+                if (localizer != null &&
+                    !string.IsNullOrEmpty(uiHintAttribute.UIHint))
+                {
+                    displayMetadata.TemplateHint = localizer[uiHintAttribute.UIHint];
+                }
+                else
+                {
+                    displayMetadata.TemplateHint = uiHintAttribute.UIHint;
+                }
```

Comment:
Why localize the `UIHint`? This string is used to find templates, not displayed to users.